src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -3566,10 +3573,18 @@ JoinOnPartitionColumn(Query *query)
 	bool joinOnPartitionColumn = false;
 	List *leftCompositeFieldList = NIL;
 	List *rightCompositeFieldList = NIL;
-	List *whereClauseList = WhereClauseList(query->jointree);
-	List *joinClauseList = JoinClauseList(whereClauseList);
-
+	List *whereClauseList = NIL;
+	List *outerJoinClauseList = NIL;
+	List *joinClauseList = NIL;
 	ListCell *joinClauseCell = NULL;
+
+	whereClauseList = WhereClauseList(query->jointree);
+
+	/* for subquery pushdown check, we also include outer join clauses */
+	outerJoinClauseList = OuterJoinClauseList(query->jointree);
+	whereClauseList = list_concat(whereClauseList, outerJoinClauseList);
+
+	joinClauseList = JoinClauseList(whereClauseList);
```

Comment:
We have different usages for JoinClauseList. Sometimes we call WhereClauseList first, then call JoinClauseList on the output of the function. Sometimes we directly call JoinClauseList on ```fromExpr->quals```.  Our fix for this bug is to determine join clauses for outer joins. Perhaps we should change the definition of JoinClauseList to use walkers and return only join clauses. This would eliminate future bugs too since whereclauselist would only return filters, and JoinClauseList would only return join clauses.